src/common/editable.js

```diff
@@ -146,47 +143,84 @@ export class Text extends Editable {
       </span>
     }
 
-    const { children } = this.props
     const { error, saving } = state
 
-    return <span>
-      <input
-        autoFocus
-        defaultValue={children}
-        onBlur={this._closeEdition}
-        onInput={this._onInput}
-        onKeyDown={this._onKeyDown}
-        readOnly={saving}
-        style={{
-          width: `${children.length + 1}ex`
-        }}
-        type='text'
-      />
+    return <span ref='mainEdition'>
+      {this._renderEdition()}
       {saving && <span>{' '}<Icon icon='loading' /></span>}
       {error != null && <span>{' '}<Icon icon='error' title={error} /></span>}
     </span>
   }
 }
 
 @propTypes({
-  defaultValue: propTypes.any,
+  children: propTypes.string,
+  value: propTypes.string.isRequired
+})
+export class Text extends Editable {
+  constructor (props) {
+    super()
+
+    this._value = props.children
+  }
+
+  _onInput = ({ target }) => {
+    target.style.width = `${target.value.length + 1}ex`
+    this._value = this.refs.input.value
+  }
+
+  _renderDisplay () {
+    const { useLongClick, children, value } = this.props
+
+    return <span className={!children && !value && 'text-muted'}>
+      {this.props.children ||
+        this.props.value ||
+        this.props.placeholder ||
+        (useLongClick ? _('editableLongClickPlaceholder') : _('editableClickPlaceholder'))
+      }
+    </span>
+  }
+
+  _renderEdition () {
+    const { children } = this.props
+    const { saving } = this.state
+
+    return <input
+      autoFocus
+      defaultValue={children}
+      onBlur={this._closeEdition}
+      onInput={this._onInput}
+      onKeyDown={this._onKeyDown}
+      readOnly={saving}
+      ref='input'
+      style={{
+        width: `${children.length + 1}ex`
+      }}
+      type='text'
+    />
+  }
+}
+
+@propTypes({
+  children: propTypes.any,
   labelProp: propTypes.string.isRequired,
-  onChange: propTypes.func.isRequired,
   options: propTypes.oneOfType([
     propTypes.array,
     propTypes.object
   ]).isRequired,
-  useLongClick: propTypes.bool
+  value: propTypes.any.isRequired
 })
 export class Select extends Editable {
   constructor (props) {
     super()
 
-    this._defaultValue = findKey(props.options, option => option === props.defaultValue)
+    this._value = props.value
+    this._defaultValue = findKey(props.options, option => option === this._value)
```

Comment:
Move into the `render()` or if you want to do things in a better way in `componentWillReceiveProps()`.